### PR TITLE
feat: Improvements to new storage interface

### PR DESCRIPTION
**Pull request by DRK3**
- Improved documentation.
- Moved CloseStore method from Provider to Store.
- Introduced custom error type for bulk methods so the caller can determine which operation failed.

Signed-off-by: Derek Trider <Derek.Trider@securekey.com>

### CODE REVIEW

File: pkg/newstorage/newstorage.go [hunk 1]
```diff
@@ -8,6 +8,12 @@ package newstorage
 
 import "errors"
 
+// BulkOperationError represents one or more errors that occurred during a bulk operation.
+type BulkOperationError interface {
+	error
```
Comment (DRK3) [inline pkg/newstorage/newstorage.go:13]: BulkOperationError.Error() would return a summary of all errors

File: pkg/newstorage/newstorage.go [hunk 2]
```diff
@@ -99,16 +109,15 @@ type Store interface {
...
 
-// Iterator represents an iterator that can be used to iterate over stored key-value pairs.
+// Iterator allows for iteration over a collection of key-value pairs.
 type Iterator interface {
 	// Next moves the pointer to the next value in the iterator. It returns false if the iterator is exhausted.
 	Next() (bool, error)
 
-	// Release releases associated resources. Release should always result in success
-	// and can be called multiple times without causing an error.
+	// Release releases associated resources.
```
Comment (DRK3) [inline pkg/newstorage/newstorage.go:120]: It was weird that the method said that it should always result in success despite it still being able to return an error. I've simplified it and removed that wording. Most of our implementations can throw errors during Release(), so it seems like we still need an error return value.

File: pkg/newstorage/newstorage.go [hunk 3]
```diff
@@ -84,9 +94,9 @@ type Store interface {
 	GetTags(k string) ([]Tag, error)
 
 	// GetBulk fetches the values associated with the given keys.
-	// If a key doesn't exist, then a nil []byte is returned for that value and is not considered an error.
+	// If a key doesn't exist, then a nil []byte is returned for that value. It is not considered an error.
 	// Depending on the implementation, this method may be faster than calling Get for each key individually.
-	GetBulk(keys ...string) ([][]byte, error)
```
Comment (troyronda) [inline pkg/newstorage/newstorage.go:89]: You should return a generic error type from functions.

File: pkg/newstorage/newstorage.go [hunk 4]
```diff
@@ -99,16 +109,15 @@ type Store interface {
 
 	// Batch performs multiple Put and/or Delete operations in order.
 	// Depending on the implementation, this method may be faster than repeated Put and/or Delete calls.
-	Batch(operations []Operation) error
```
Comment (troyronda) [inline pkg/newstorage/newstorage.go:102]: You should return a generic error type from functions.

File: pkg/newstorage/newstorage.go [hunk 5]
```diff
@@ -8,6 +8,12 @@ package newstorage
 
 import "errors"
 
+// BulkOperationError represents one or more errors that occurred during a bulk operation.
+type BulkOperationError interface {
+	error
+	Errors() map[int]error // Errors returns a mapping of failed operation indices to their associated errors.
```
Comment (troyronda) [inline pkg/newstorage/newstorage.go:14]: why a map[int] instead of a slice?

File: pkg/newstorage/newstorage.go [hunk 6]
```diff
@@ -8,6 +8,12 @@ package newstorage
 
 import "errors"
 
+// BulkOperationError represents one or more errors that occurred during a bulk operation.
+type BulkOperationError interface {
```
Comment (troyronda) [inline pkg/newstorage/newstorage.go:12]: also this pattern is more generally the multierror.

File: pkg/newstorage/newstorage.go (hunk 5 above)
Comment (DRK3) [reply pkg/newstorage/newstorage.go:14]: My thinking is that a caller might want the ability to know exactly which operation failed.

**Comment by troyronda** [comment]
We also need to decide what to do with CloseStore (from prior comments).

File: pkg/newstorage/newstorage.go (hunk 5 above)
Comment (DRK3) [reply pkg/newstorage/newstorage.go]: Changed to slice.

File: pkg/newstorage/newstorage.go (hunk 6 above)
Comment (DRK3) [reply pkg/newstorage/newstorage.go]: Changed to MultiError.

**Comment by DRK3** [comment]
> We also need to decide what to do with CloseStore (from prior comments).

I've moved it to the Store interface after some discussion

**Review by troyronda — APPROVED** [review]

**Pull request merged**